relayer/chains/solana/listener.go

```diff
@@ -28,19 +29,222 @@ func (p *Provider) Listener(ctx context.Context, lastSavedHeight uint64, blockIn
 	// 	p.log.Error("failed to init xcall", zap.Error(err))
 	// }
 
+	// return nil
+
 	if err := p.SendMessage(ctx, &relayertypes.Message{
 		Dst:  "0x3.icon",
 		Data: []byte("hello"),
 	}); err != nil {
 		p.log.Error("failed to send message", zap.Error(err))
 	}
 
-	p.log.Info("started querying from height", zap.Uint64("height", startHeight))
+	// fromSignature := "2D1RxZptfGrfHwqfzMD3Z6TdnGb5Ugff3ZHuDmNT7xMTfpRtaBSpdwrC2R8qrioDDFvkU3TU5yTSukv2iByoGcuN"
+	fromSignature := ""
+
+	p.log.Info("started querying from height", zap.String("from-signature", fromSignature))
+
+	return p.listenByPolling(ctx, fromSignature, blockInfo)
+}
+
+func (p *Provider) listenByPolling(ctx context.Context, fromSignature string, blockInfo chan *relayertypes.BlockInfo) error {
+	ticker := time.NewTicker(3 * time.Second)
+
+	if fromSignature != "" {
+		fromSign, err := solana.SignatureFromBase58(fromSignature)
+		if err != nil {
+			return err
+		}
+		if err := p.processTxSignature(ctx, fromSign, blockInfo); err != nil {
+			p.log.Error("failed to process tx signature", zap.String("signature", fromSign.String()), zap.Error(err))
+		}
+	}
 
 	for {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
+		case <-ticker.C:
+			//fetch txSigns from most recent to oldest. 0th index is the most recent and last index is oldest
+			txSigns, err := p.getSignatures(ctx, fromSignature)
+			if err != nil {
+				p.log.Error("failed to get signatures", zap.Error(err))
+				break
+			}
+			if len(txSigns) > 0 {
+				//next query start from most recent signature.
+				fromSignature = txSigns[0].Signature.String()
```

Comment:
any reason the method parameter is reset here ?

Reply:
In that case, can we make this a variable in the provider itself rather than changing the function parameter ?

Reply:
We could but I am not able to see its benefits. Keeping it in provider type would result in a shared state. However, based on its usecase we don't need that shared state, it has only local scope which is while fetching signatures only.